### PR TITLE
Fix objPoints order in GridBoard and CharucoBoard

### PR DESCRIPTION
**Pull request by AleksandrPanov**
Fixes https://github.com/opencv/opencv_contrib/issues/2623 and https://github.com/opencv/opencv_contrib/issues/2604

This PR:

**Fixed objPoints order in GridBoard and CharucoBoard**.
In the `_drawPlanarBoardImpl()` function the y coordiantes of the charuco markers get horizontally mirrored by `pf.y = (1.0f - pf.y / sizeY) * float(out.rows);` Now `pf.y = pf.y / sizeY * float(out.rows);`

Drawing correctness was checked by [example.cpp.txt](https://github.com/opencv/opencv_contrib/files/8074445/example.cpp.txt)

**Before fix:**
- Arrays `objPoints` and `ids` for 3x4 `aruco::GridBoard` are in the wrong order
board_detected_grid
- Arrays `objPoints` and `ids` for 3x4 `aruco::CharucoBoard` are in the wrong order. Also `objPoints` coordinates are incorrect.
charuco_detected_grid

**After fix:**
image
Everything is drawn correctly.

**Fixed generation of rotated images (in ArUco tests). Now rotated images are generated using the "[aircraft principal axes](https://en.wikipedia.org/wiki/Aircraft_principal_axes)".**

image
rotate_pitch
rotate_yaw

**Also fixed `objPoints[markerIndex]` order (from CCW to CW) in GridBoard and CharucoBoard.** Because tutorial and all ArUco algs use CW order.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/aruco/include/opencv2/aruco.hpp [hunk 1]
```diff
@@ -303,6 +303,8 @@ class CV_EXPORTS_W Board {
     /// vector of the identifiers of the markers in the board (same size than objPoints)
     /// The identifiers refers to the board dictionary
     CV_PROP_RW std::vector< int > ids;
+
+    CV_PROP Point3f rightBottomBorder;
```
Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco.hpp:307]: Please add documentation line for the property.

File: modules/aruco/test/test_aruco_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+         * @brief Get rvec and tvec from yaw, pitch and distance
+         */
+        static inline void getSyntheticRT(double yaw, double pitch, double distance, Mat& rvec, Mat& tvec) {
+
+            rvec = Mat::zeros(3, 1, CV_64FC1);
+            tvec = Mat::zeros(3, 1, CV_64FC1);
+
+            // rotate camera in pitch axis
+            Mat rotPitch(3, 1, CV_64FC1);
+            rotPitch.ptr< double >(0)[0] = pitch;
```
Comment (asmorkalov) [inline modules/aruco/test/test_aruco_utils.hpp:17]: rotPitch.at< double >(i, j) is better.

**Comment by AleksandrPanov** [comment]
rotate_yaw

File: modules/aruco/include/opencv2/aruco.hpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco.hpp:308]: added

File: modules/aruco/test/test_aruco_utils.hpp (hunk 2 above)
Comment (AleksandrPanov) [reply modules/aruco/test/test_aruco_utils.hpp]: fixed

File: modules/aruco/include/opencv2/aruco.hpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco.hpp]: image

**Review by asmorkalov — APPROVED** [review]

File: modules/aruco/include/opencv2/aruco.hpp [hunk 3]
```diff
@@ -294,7 +294,7 @@ class CV_EXPORTS_W Board {
     CV_WRAP void setIds(InputArray ids);
 
     /// array of object points of all the marker corners in the board
-    /// each marker include its 4 corners in CCW order. For M markers, the size is Mx4.
+    /// each marker include its 4 corners in CW order. For M markers, the size is Mx4.
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco.hpp:297]: It doesn't make sense to mention CW or CCW without specifying used coordinate system.

Please take a look on calib3d module documentation.

File: modules/aruco/test/test_aruco_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
+#include "test_precomp.hpp"
```
Comment (alalek) [inline modules/aruco/test/test_aruco_utils.hpp:1]: Missing license header

File: modules/aruco/src/aruco.cpp [hunk 5]
```diff
@@ -1713,20 +1719,19 @@ Ptr<GridBoard> GridBoard::create(int markersX, int markersY, float markerLength,
...
-            corners.resize(4);
+            vector<Point3f> corners(4);
             corners[0] = Point3f(x * (markerLength + markerSeparation),
-                                 maxY - y * (markerLength + markerSeparation), 0);
+                                 y * (markerLength + markerSeparation), 0);
             corners[1] = corners[0] + Point3f(markerLength, 0, 0);
-            corners[2] = corners[0] + Point3f(markerLength, -markerLength, 0);
-            corners[3] = corners[0] + Point3f(0, -markerLength, 0);
+            corners[2] = corners[0] + Point3f(markerLength, markerLength, 0);
+            corners[3] = corners[0] + Point3f(0, markerLength, 0);
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp:1729]: You just changed the coordinate system by flipping Y.

File: modules/aruco/test/test_aruco_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+namespace opencv_test {
+    namespace {
+
+        static inline double deg2rad(double deg) { return deg * CV_PI / 180.; }
```
Comment (alalek) [inline modules/aruco/test/test_aruco_utils.hpp:5]: avoid indentation in namespaces

File: modules/aruco/src/aruco.cpp (hunk 5 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp:1729]: This changes was added to make coordinate systems consistent.
Now in `Board::create()` use this coordinate system:
image
But `Board::draw()` use another coordinate system:
image

Also, `Board::create()` set corners for marker in `objPoints` with "CCW" order:
image
But `detectMarkers()` find markers and reordered corners to "CW" order:
image

This makes coordinate systems not consistent and makes the code hard to read

File: modules/aruco/test/test_aruco_utils.hpp (hunk 6 above)
Comment (AleksandrPanov) [reply modules/aruco/test/test_aruco_utils.hpp]: fixed

File: modules/aruco/test/test_aruco_utils.hpp (hunk 4 above)
Comment (AleksandrPanov) [reply modules/aruco/test/test_aruco_utils.hpp:4]: fixed

File: modules/aruco/include/opencv2/aruco.hpp (hunk 3 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco.hpp]: Added this definition:
```
    /// each marker include its 4 corners in this order:
    ///-   objPoints[i][0] - left-top point of i-th marker
    ///-   objPoints[i][1] - right-top point of i-th marker
    ///-   objPoints[i][2] - right-bottom point of i-th marker
    ///-   objPoints[i][3] - left-bottom point of i-th marker
```

File: modules/aruco/test/test_aruco_utils.hpp (hunk 6 above)
Comment (AleksandrPanov) [reply modules/aruco/test/test_aruco_utils.hpp]: @alalek thanks for the review)
do you have any more questions?

**Comment by AleksandrPanov** [comment]
@alalek thanks for the review)
do you have any more questions?

File: modules/aruco/src/aruco.cpp [hunk 7]
```diff
@@ -1782,9 +1787,18 @@ void drawDetectedMarkers(InputOutputArray _image, InputArrayOfArrays _corners,
...
 void drawAxis(InputOutputArray _image, InputArray _cameraMatrix, InputArray _distCoeffs, InputArray _rvec,
-              InputArray _tvec, float length)
+              InputArray _tvec, float length, int thickness)
 {
-    drawFrameAxes(_image, _cameraMatrix, _distCoeffs, _rvec, _tvec, length, 3);
+    vector<Point3f> axis;
+    axis.push_back(Point3f(0.f, 0.f, 0.f));
+    axis.push_back(Point3f(length, 0.f, 0.f));
+    axis.push_back(Point3f(0.f, length, 0.f));
+    axis.push_back(Point3f(0.f, 0.f, -length));
```
Comment (asmorkalov) [inline modules/aruco/src/aruco.cpp:1796]: why `-length`? Your function differs from standard drawFrameAxes with opposite direction of the third axis.

File: modules/aruco/test/test_aruco_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+static inline vector<Point2f> getAxis(InputArray _cameraMatrix, InputArray _distCoeffs,
+                                      InputArray _rvec, InputArray _tvec, float length,
+                                            const float offset = 0.f)
+{
+    vector<Point3f> axis;
+    axis.push_back(Point3f(offset, offset, 0.f));
+    axis.push_back(Point3f(length+offset, offset, 0.f));
+    axis.push_back(Point3f(offset, length+offset, 0.f));
+    axis.push_back(Point3f(offset, offset, -length));
```
Comment (asmorkalov) [inline modules/aruco/test/test_aruco_utils.hpp:16]: The same, the result axis direction is opposite to what standard drawFrameAxes does.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/aruco/src/aruco.cpp (hunk 7 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp]: This was added to improve appearance. The drawAxis() function doesn't make sense, as it is just calls drawFrameAxes().

**Comment by AleksandrPanov** [comment]
I checked the coordinate system axis from pose estimation:

Chessboard pose estimation:
image

ArUco board pose estimation (after fix):
image

ChArUco board pose estimation (after fix):
image

The `drawAxis()` function doesn't make sense, as it is just calls `drawFrameAxes()` and because `drawAxis()` was removed.

Also `drawCustomAxis()` was added:
image

[poseEstimate.py.txt](https://github.com/opencv/opencv_contrib/files/8181631/poseEstimate.py.txt)
Use this image - aruco\tutorials\aruco_calibration\images\img_00.jpg:
img_00

File: modules/aruco/test/test_aruco_utils.hpp (hunk 8 above)
Comment (AleksandrPanov) [reply modules/aruco/test/test_aruco_utils.hpp]: fixed

**Review by asmorkalov — APPROVED** [review]

**Review by alalek — APPROVED** [review]

**Comment by catree** [comment]
First very quick look.

---

Not sure to understand the purpose of `cv::aruco::drawCustomAxis()` vs `cv::drawFrameAxes()`:
- `void drawCustomAxis(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3, Point3f center = Point3f(0.f, 0.f, 0.f), Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f))`
- `void drawFrameAxes(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3)`

About the shifted center point `Point3f center = Point3f(0.f, 0.f, 0.f)`, I think what you are doing is:

image

But this is equivalent to:

image

About the `Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f)`, is there really a need to use different sizes for the frame axes?

```python
cv.aruco.drawCustomAxis(img, K, D, rvecs, tvecs, 1., 3, (0, 7, 0), (1, -1, -1))
```

Should be equivalent to:

image

image

**Comment by AleksandrPanov** [comment]
> First very quick look.
> 
> Not sure to understand the purpose of `cv::aruco::drawCustomAxis()` vs `cv::drawFrameAxes()`:
> 
> * `void drawCustomAxis(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3, Point3f center = Point3f(0.f, 0.f, 0.f), Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f))`
> * `void drawFrameAxes(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3)`
> 
> About the shifted center point `Point3f center = Point3f(0.f, 0.f, 0.f)`, I think what you are doing is:
> 
> image
> 
> But this is equivalent to:
> 
> image
> 
> About the `Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f)`, is there really a need to use different sizes for the frame axes?
> 
> ```python
> cv.aruco.drawCustomAxis(img, K, D, rvecs, tvecs, 1., 3, (0, 7, 0), (1, -1, -1))
> ```
> 
> Should be equivalent to:
> 
> image
> 
> image

@catree the direction/position of the axes changed after objPoints fix. A `drawCustomAxis()` has been added to draw the axes in the old style to avoid reworking the tutorial. Do I understand correctly that you are proposing to use transformations with rvecs, tvecs vectors to get the desired axis direction/position?

**Comment by catree** [comment]
I am not fan of the `drawCustomAxis()` because it feels unnecessary for me. I would prefer some good API for handling SE3/SO3 transformations with some good tutorials. You can also shoot yourself in the foot if you use `axisCoefficient=(-1,0,0)` or any other non valid values.

What about [`composeRT()`](https://docs.opencv.org/4.5.5/d9/d0c/group__calib3d.html#ga044b03d81f99370db05c52c3d0b46599), can it do the same job?

**Comment by AleksandrPanov** [comment]
> I am not fan of the `drawCustomAxis()` because it feels unnecessary for me. I would prefer some good API for handling SE3/SO3 transformations with some good tutorials. You can also shoot yourself in the foot if you use `axisCoefficient=(-1,0,0)` or any other non valid values.
> 
> Moreover, the extra functionality of `drawCustomAxis()` is not used in your PR (only for debugging purposes), no?
> 
> What about [`composeRT()`](https://docs.opencv.org/4.5.5/d9/d0c/group__calib3d.html#ga044b03d81f99370db05c52c3d0b46599), can it do the same job?

@catree, thx)
I removed `drawCustomAxis()` and added `cv::drawFrameAxes()` link to docs and replaced old images with new ones. The tutorial now shows the correct direction and position of the axes.

**Comment by AleksandrPanov** [comment]
@catree, I think that good API for handling SE3/SO3 transformations should add into another PR. This PR fixes only ArUco problems.

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Comment by mabl** [comment]
It looks like this change makes `interpolateCornersCharuco`  return a different number for board corners. [See image here](https://github.com/opencv/opencv_contrib/commit/56d492cfecb7aadaca79a9494a33ad1c12798e5e#diff-de110295d0b9a89eb14a5aedc00a241012295be05e4ab60776f3ec94f5227058). Before this, this corner id mapped to the index of `board->chessboardCorners`. 

Is this change intentional? If so, how can I recover the corresponding index for `board->chessboardCorners` from [interpolateCornersCharuco](https://docs.opencv.org/3.4/d9/d6a/group__aruco.html#gadcc5dc30c9ad33dcf839e84e8638dcd1) markerIds, i.e. the list of identifiers for each marker in corners.

**Pull request merged**